src/macaron/repo_finder/repo_validator.py

```diff
@@ -30,3 +43,14 @@ def find_valid_repository_url(urls: Iterable[str]) -> str:
 
     # Report the first valid URL from the end of the list.
     return vcs_list.pop()
+
+
+def _resolve_redirects(parsed_url: urllib.parse.ParseResult) -> str | None:
```

Comment:
Because we are trying to resolve the final redirected URL first before passing it to `get_remote_vcs_url`, I wonder why don't we let `requests` handles the redirection for us (as mentioned [here](https://requests.readthedocs.io/en/latest/user/quickstart/?highlight=allow_redirects#redirection-and-history)) and only care about the final URL that it lands on. In the current implementation, we only stops at the first redirection, I'm curious if there would be any case for multiple redirection in our dataset of repositories :thinking: .

Reply:
Right, so we only supports URL with at most 1 redirection. Then I think we can leave the current implementation as it is. 
I agree that following the redirection could potentially add more cost to the operation.

Reply:
Another consideration is the security aspect, that I'll add here for the sake of completeness. Currently Macaron will only try to resolve URLs within the allowed domain list during that part of the analysis. Allowing `requests` to follow these redirects indiscriminately creates a possibility that some undesirable URL is retrieved. That said, as nothing is done with the retrieved data the threat there may still be minimal.